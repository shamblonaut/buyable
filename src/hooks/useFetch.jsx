import { useEffect, useState } from "react";

const useFetch = (url) => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetch(url)
      .then((response) => {
        if (!response.ok) {
          throw new Error(
            `Error ${response.status}${response.statusText ? `: ${response.statusText}` : ""}`,
          );
        }

        return response.json();
      })
      .then((response) => setData(response))
      .catch((error) => setError(error.message))
      .finally(() => setLoading(false));
  }, [url]);

  return { data, error, loading };
};

export default useFetch;
