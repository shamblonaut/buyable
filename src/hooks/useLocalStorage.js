import { useEffect, useState } from "react";

const useLocalStorage = (name, initialValue = null) => {
  const [value, setValue] = useState(() => {
    const storedValue = localStorage.getItem(name);
    return storedValue !== null ? JSON.parse(storedValue) : initialValue;
  });

  useEffect(() => {
    localStorage.setItem(name, JSON.stringify(value));
  }, [name, value]);

  return [
    value,
    (newValue) =>
      setValue((currentValue) =>
        typeof newValue === "function" ? newValue(currentValue) : newValue,
      ),
  ];
};

export default useLocalStorage;
