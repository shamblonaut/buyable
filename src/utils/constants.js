export const AppPosition = Object.freeze({
  HOME: 0,
  SHOP: 1,
  CART: 2,
});
