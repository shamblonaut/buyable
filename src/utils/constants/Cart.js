export const CartAction = Object.freeze({
  ADDED: "added_to_cart",
  REMOVED: "removed_from_cart",
  INCREMENTED: "incremented_cart_quantity",
  DECREMENTED: "decremented_cart_quantity",
  EDITED: "edited_cart_quantity",
  CLEARED: "cleared_cart",
});
