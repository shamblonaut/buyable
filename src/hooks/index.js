import useFetch from "./useFetch/useFetch";
import useLocalStorage from "./useLocalStorage/useLocalStorage";
import useCart from "./useCart/useCart";

export { useFetch, useLocalStorage, useCart };
