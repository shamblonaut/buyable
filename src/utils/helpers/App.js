import { AppPosition } from "@/utils/constants";

export const getPositionRoute = (appPosition) => {
  switch (appPosition) {
    case AppPosition.HOME:
      return "/";
    case AppPosition.SHOP:
      return "/shop";
    case AppPosition.CART:
      return "/cart";
  }
};
