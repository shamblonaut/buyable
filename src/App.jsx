import { useState } from "react";
import { Outlet } from "react-router-dom";

import { AppPosition } from "@/utils/constants";

import { CartContext } from "@/contexts";
import { useFetch, useLocalStorage } from "@/hooks";

import { Header, Navigator } from "@/components";
import { RootStyle } from "@/styles";

const App = () => {
  const productsData = useFetch("https://fakestoreapi.com/products");
  const [cart, setCart] = useLocalStorage("cart", {});

  const [appPosition, setAppPosition] = useState(AppPosition.HOME);

  return (
    <CartContext.Provider value={{ cart, setCart }}>
      <RootStyle />
      <Header />
      <Outlet
        context={{
          productsData,
          appPosition,
          setAppPosition,
        }}
      />
      <Navigator appPosition={appPosition} />
    </CartContext.Provider>
  );
};

export default App;
