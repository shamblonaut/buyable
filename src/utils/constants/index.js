import { AppPosition } from "./App";

export { AppPosition };
