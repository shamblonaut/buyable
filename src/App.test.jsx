import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";

import { BrowserRouter } from "react-router-dom";

import App from "./App";

import { AppPosition } from "@/utils/constants";
import { useFetch, useLocalStorage } from "@/hooks";

vi.mock("react-router-dom", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    Outlet: () => <div data-testid="outlet-page">Mocked Page</div>,
  };
});

vi.mock("@/hooks", () => ({
  useFetch: vi.fn(),
  useLocalStorage: vi.fn(),
}));

vi.mock("@/components", () => ({
  Header: () => <header data-testid="header">BUYABLE</header>,
  Navigator: ({ appPosition, cartCount }) => (
    <nav data-testid="navigator">
      <a>Home{appPosition === AppPosition.HOME && <> [active]</>}</a>
      <a>Shop{appPosition === AppPosition.SHOP && <> [active]</>}</a>
      <a>
        Cart: {cartCount} items
        {appPosition === AppPosition.CART && <> [active]</>}
      </a>
    </nav>
  ),
}));

describe("App", () => {
  beforeEach(() => {
    vi.clearAllMocks();

    useFetch.mockReturnValue({
      data: null,
      error: null,
      loading: true,
    });
    useLocalStorage.mockReturnValue([{ 1: 6, 2: 4 }, vi.fn()]);
  });

  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByRole("banner")).toBeInTheDocument(); // header
    expect(screen.getByRole("navigation")).toBeInTheDocument(); // nav
  });

  it("activates home tab by default", () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    );

    expect(screen.getByText("Home [active]")).toBeInTheDocument();
  });

  it("passes correct amount for number of items in cart", () => {
    useLocalStorage.mockReturnValue([{ 1: 8, 2: 3, 3: 9 }, vi.fn()]);

    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    );

    expect(screen.getByText("Cart: 20 items")).toBeInTheDocument();
  });
});
