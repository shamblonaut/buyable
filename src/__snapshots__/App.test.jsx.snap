// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders correctly 1`] = `
<div>
  <header
    data-testid="header"
  >
    BUYABLE
  </header>
  <div
    data-testid="outlet-page"
  >
    Mocked Page
  </div>
  <nav
    data-testid="navigator"
  >
    <a>
      Home
       [active]
    </a>
    <a>
      Shop
    </a>
    <a>
      Cart: 
      10
       items
    </a>
  </nav>
</div>
`;
